File: yep/src/main/java/catchla/yep/view/VoiceWaveView.java
package catchla.yep.view;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.view.View;

import org.apache.commons.lang3.ArrayUtils;

import java.util.ArrayList;

import catchla.yep.R;
import catchla.yep.util.MathUtils;

/**
 * Created by kevin on 14/12/23.
 */
public class VoiceWaveView extends View {

    public int mAmplitude;
    private Path mPath;
    private float mDimension;
    private float mDimension1;
    private int numberOfWaves = 5;
    private float mFrequency = 1.2f;
    private float density = 1.f;
    private double mPhase = 0.f;
    private float mPhaseShift = 0.25f;
    private float mMaxAmplitude = 0;
    private boolean drawlock = false;
    private Paint paintsArray;
    private int mViewWidth = 0;
    private int mViewHeight = 0;
    private float ViewMid = 0;
    private boolean mRecordingStarted;
    private ArrayList<Float> mSamplesList = new ArrayList<>();

    public VoiceWaveView(final Context context, final AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public VoiceWaveView(Context context) {
        super(context);
        init();
    }

    public int getAmplitude() {
        return mAmplitude;
    }

    public void setAmplitude(final int amplitude) {
        mAmplitude = (mAmplitude + amplitude) / 2;
        mSamplesList.add(amplitude / (float) Short.MAX_VALUE);
        phaseNext();
        invalidate();
    }

    private void init() {
        Resources res = getResources();
        paintsArray = new Paint(Paint.ANTI_ALIAS_FLAG);
        paintsArray.setColor(0x32a7ff);
        paintsArray.setStyle(Paint.Style.STROKE);

        mDimension = res.getDimension(R.dimen.waver_width);
        mDimension1 = res.getDimension(R.dimen.waver_width_min);

        mPath = new Path();

        setAmplitude(0);
    }


    @Override
    protected void onSizeChanged(int xNew, int yNew, int xOld, int yOld) {
        super.onSizeChanged(xNew, yNew, xOld, yOld);
        mViewWidth = xNew;
        mViewHeight = yNew;
        ViewMid = mViewWidth / 2.0f;
        mMaxAmplitude = mViewHeight / 2.f - 4.0f;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (drawlock) {
            return;
        } else {
            drawlock = true;
        }
        for (int i = 0; i < numberOfWaves; i++) {

            float progress = 1.f - (i / (float) numberOfWaves);
            float normedAmplitude = (1.5f * progress - 0.5f) * mAmplitude / 65536f;

            mPath.reset();

            for (float x = 0.f; x < mViewWidth + density; x += density) {

                //Thanks to https://github.com/stefanceriu/SCSiriWaveformView
                // We use a parable to scale the sinus wave, that has its peak in the middle of the view.
                double scaling = -Math.pow(x / ViewMid - 1.f, 2.f) + 1.f; // make center bigger

                double y = scaling * mMaxAmplitude * normedAmplitude * Math.sin(2 * Math.PI * (x / mViewWidth)
                        * mFrequency + mPhase) + mViewHeight / 2.0;

                if (x == 0.f) {
                    mPath.moveTo(x, (float) y);
                } else {
                    mPath.lineTo(x, (float) y);
                }
            }
            float multiplier = Math.min(1.0f, (progress / 3.0f * 2.0f) + (1.0f / 3.0f));

            if (i == 0) {
                paintsArray.setAlpha(255);
                paintsArray.setStrokeWidth(mDimension);
            } else {
                paintsArray.setAlpha((int) ((1.0f * multiplier * 0.4) * 255));
                paintsArray.setStrokeWidth(mDimension1);
            }
            canvas.drawPath(mPath, paintsArray);

        }

        drawlock = false;

    }

    private double phaseNext() {
        mPhase += mPhaseShift;
        if (mPhase > 2 * Math.PI) {
            mPhase = mPhase - 2 * Math.PI;
        }
        return mPhase;
    }


    public float[] stopRecording() {
        mRecordingStarted = false;
        final int size = mSamplesList.size();
        final float[] rawSamplesArray = ArrayUtils.toPrimitive(mSamplesList.toArray(new Float[size]));
        final int idealSampleSize = 20;
        if (size < idealSampleSize) {
            return rawSamplesArray;
        }
        final int gap = size / idealSampleSize;
        final float[] result = new float[idealSampleSize];
        for (int i = 0; i < idealSampleSize; i++) {
            result[i] = MathUtils.avg(rawSamplesArray, i * gap, (i + 1) * gap - 1);
        }
        return result;
    }

    public void startRecording() {
        mRecordingStarted = true;
        mSamplesList.clear();
    }
}

File: src/main/kotlin/catchla/yep/view/VoiceWaveView.kt
package catchla.yep.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.util.AttributeSet
import android.view.View
import catchla.yep.R

/**
 * Created by kevin on 14/12/23.
 */
class VoiceWaveView : View {

    var amplitude: Int = 0
        set(value) {
            field = (field + value) / 2
            phaseNext()
            invalidate()
        }
    private val path = Path()
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)

    private var mDimension: Float = 0.toFloat()
    private var mDimension1: Float = 0.toFloat()
    private val numberOfWaves = 5
    private val mFrequency = 1.2f
    private val density = 1.0f
    private var phase = 0.0
    private val mPhaseShift = 0.25f
    private var maxAmplitude = 0f
    private var drawlock = false
    private var viewWidth = 0
    private var viewHeight = 0
    private var viewMid = 0f
    private var recordingStarted: Boolean = false

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context) : super(context) {
        init()
    }

    private fun init() {
        val res = resources
        paint.color = 0x32a7ff
        paint.style = Paint.Style.STROKE

        mDimension = res.getDimension(R.dimen.waver_width)
        mDimension1 = res.getDimension(R.dimen.waver_width_min)

        amplitude = 0
    }


    override fun onSizeChanged(xNew: Int, yNew: Int, xOld: Int, yOld: Int) {
        super.onSizeChanged(xNew, yNew, xOld, yOld)
        viewWidth = xNew
        viewHeight = yNew
        viewMid = viewWidth / 2.0f
        maxAmplitude = viewHeight / 2.0f - 4.0f
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (drawlock) {
            return
        } else {
            drawlock = true
        }
        for (i in 0 until numberOfWaves) {

            val progress = 1.0f - i / numberOfWaves.toFloat()
            val normedAmplitude = (1.5f * progress - 0.5f) * amplitude / 65536f

            path.reset()

            var x = 0.0f
            while (x < viewWidth + density) {

                //Thanks to https://github.com/stefanceriu/SCSiriWaveformView
                // We use a parable to scale the sinus wave, that has its peak in the middle of the view.
                val scaling = -Math.pow((x / viewMid - 1.0f).toDouble(), 2.0) + 1.0f // make center bigger

                val y = scaling * maxAmplitude.toDouble() * normedAmplitude.toDouble() * Math.sin(2.0 * Math.PI * (x / viewWidth).toDouble() * mFrequency.toDouble() + phase) + viewHeight / 2.0

                if (x == 0.0f) {
                    path.moveTo(x, y.toFloat())
                } else {
                    path.lineTo(x, y.toFloat())
                }
                x += density
            }
            val multiplier = Math.min(1.0f, progress / 3.0f * 2.0f + 1.0f / 3.0f)

            if (i == 0) {
                paint.alpha = 255
                paint.strokeWidth = mDimension
            } else {
                paint.alpha = (1.0 * multiplier.toDouble() * 0.4 * 255).toInt()
                paint.strokeWidth = mDimension1
            }
            canvas.drawPath(path, paint)

        }

        drawlock = false

    }

    private fun phaseNext(): Double {
        phase += mPhaseShift.toDouble()
        if (phase > 2 * Math.PI) {
            phase -= 2 * Math.PI
        }
        return phase
    }


    fun stopRecording() {
        recordingStarted = false
    }

    fun startRecording() {
        recordingStarted = true
    }
}